ui/apps/platform/src/services/DiscoveredClusterService.ts

```diff
@@ -0,0 +1,214 @@
+        .then((response) => response.data.count);
+}
+
+export type GetDiscoveredClusterResponse = {
+    cluster: DiscoveredCluster;
+};
+
+export function getDiscoveredCluster(id: string): Promise<DiscoveredCluster> {
+    return axios
+        .get<GetDiscoveredClusterResponse>(`${basePath}/${id}`)
+        .then((response) => response.data.cluster);
+}
+
+export type ListDiscoveredClustersRequest = {
+    filter: DiscoveredClustersFilter;
+    pagination: Pagination;
+};
+
+export type ListDiscoveredClustersResponse = {
+    clusters: DiscoveredCluster[];
+};
+
+export function listDiscoveredClusters(
+    arg: ListDiscoveredClustersRequest
+): Promise<DiscoveredCluster[]> {
+    const params = qs.stringify(arg, { arrayFormat: 'repeat', allowDots: true });
+    return axios
+        .get<ListDiscoveredClustersResponse>(`${basePath}?${params}`)
+        .then((response) => response.data.clusters);
+}
+
+export function getListDiscoveredClustersArg({
+    page,
+    perPage,
+    searchFilter,
+    sortOption,
+}): ListDiscoveredClustersRequest {
+    const filter = getDiscoveredClustersFilter(searchFilter);
+    const pagination: Pagination = { limit: perPage, offset: page - 1, sortOption };
+    return { filter, pagination };
+}
+
+export type DiscoveredClustersFilter = {
+    names?: string[];
+    sourceIds?: string[];
+    statuses?: DiscoveredClusterStatus[];
+    types?: DiscoveredClusterType[];
+};
+
+// Given searchFilter from useURLSearch hook, return validated filter argument.
+
+// See proto/storage/discovered_cluster.proto
+export const nameField = 'Cluster'; // metadata.name is for filter and sort
+export const sourceIdField = 'Integration ID'; // source.id
+export const statusField = 'Cluster Status'; // status
+export const typeField = 'Cluster Type'; // metadata.type
+
+export function getDiscoveredClustersFilter(searchFilter: SearchFilter): DiscoveredClustersFilter {
+    return {
+        names: getValues(searchFilter[nameField]),
+        sourceIds: getValues(searchFilter[sourceIdField]),
+        statuses: getStatuses(searchFilter[statusField]),
+        types: getTypes(searchFilter[typeField]),
+    };
+}
+
+export function hasDiscoveredClustersFilter(searchFilter: SearchFilter) {
+    const { names, sourceIds, statuses, types } = getDiscoveredClustersFilter(searchFilter);
+    return hasItems(names) || hasItems(sourceIds) || hasItems(statuses) || hasItems(types);
+}
+
+function hasItems(arg: unknown[] | undefined) {
+    return Array.isArray(arg) && arg.length !== 0;
+}
+
+type SearchFilterValue = string | string[] | undefined;
+
+// names and sourceIds
+
+function getValues(arg: SearchFilterValue): string[] | undefined {
+    if (typeof arg === 'string') {
+        return [arg];
+    }
+
+    if (Array.isArray(arg)) {
+        return arg;
+    }
+
+    return undefined;
+}
+
+// statuses
+
+function getStatuses(arg: SearchFilterValue): DiscoveredClusterStatus[] | undefined {
```

Comment:
We have a similar function in `searchUtils.ts` (`searchValueAsArray`) that does the same thing with the following exceptions:
- It defaults to an empty array instead of `undefined`
- It doesn't accept a type predicate to narrow a wider `string[]` argument

Depending on the semantics of how we use `undefined` vs `[]`, we might be able to generalize these into a single function in the future.